src/book/02-system/01-nubots/05-vision.mdx

```diff
@@ -5,3 +5,94 @@ title: Vision
+
+After clustering has finished, one cluster will remain which should represent the field. An upper convex hull algorithm is applied to the final cluster to determine the edge of the field. The detector will emit a GreenHorizon message that specifies the location of the edge of the field.
+
+## Ball Detector
+
+The [ball detector receives](https://github.com/NUbots/NUbots/blob/master/module/vision/BallDetector/) a GreenHorizon message and a [FieldDescription message](https://github.com/NUbots/NUbots/blob/master/shared/message/support/FieldDescription.proto). The FieldDescription message is output by the [SoccerConfig module](https://github.com/NUbots/NUbots/tree/master/module/support/configuration/SoccerConfig) which creates the FieldDescription message from configuration values specifying the layout of the soccer field.
+
+The ball detector forms clusters out of all the points that the Visual Mesh has identified as ball points. Specifically, the clusters are formed from Visual Mesh points that are identified as being a ball point, but have at least $1$ neighbour which is **not** a ball point. This allows us to form clusters of ball edge points.
+
+Any clusters which are above the field edge (i.e. the green horizon) or are smaller than a given threshold are discarded. A circular cone is then fitted to each cluster. The cone axis is determined from the line segment between the centre on the camera and the average of all ball edge points. The radius of the cone is determined by the maximum distance between the ball edge points and the average of all of the ball edge points.
+
+Degree of circle fit is used to determine if the cluster fits the shape of a circle well enough to be a ball. If it does not, then we discard it. Any ball clusters that are too close, as defined by some configuration variable, are also discarded. Clusters with a large difference between their angular and projection based distances are discarded. Any balls further away than the length of the field, as specified by the FieldDescription, are discarded.
+
+Any clusters that are still remaining are determined to be balls and so are emitted as a [Balls message](https://github.com/NUbots/NUbots/blob/master/shared/message/vision/Ball.proto).
+
+## Goal Detector
+
+The [goal detector](https://github.com/NUbots/NUbots/blob/master/module/vision/GoalDetector/src/GoalDetector.cpp) follows a similar structure to the ball detector. It receives a GreenHorizon and a FieldDescription message.
+
+Any points that are not goal points or are completely surrounded by goal points are discarded. Any points above the green horizon are also discarded. The points we have left are goal post edge points. These points are clustered. Any clusters that do not intersect the green horizon are discarded, since we assume goals are higher than the field. Any clusters that are smaller than a given threshold are also discarded.
```

Comment:
```suggestion
All goal post edge points (point that are goal points that have at least one neighbour that is **not** a goal post) are found and partitioned into clusters. Any clusters that do not intersect the green horizon are discarded, since we assume goals are higher than the field. Any clusters that are smaller than a given threshold are also discarded.
```